### PR TITLE
Add HMQV implementation (and merge the old FHMQV into the new codebase)

### PR DESCRIPTION
**Pull request by mouse07410**
Merges FHMQV implementation that existed as an unsupported patch, together with test-vectors and validation suite.

Adds HMQV implementation following the [HMQV paper](http://eprint.iacr.org/2005/176), together with test-vectors and validation suite.

### CODE REVIEW

**Comment by noloader** [comment]
Very nice, indeed!!! Let's get it in front of the group.

A couple quick comments...

* add files to FileList.txt.
* change attribution to you and I
* state copyright is assigned to project

**Comment by mouse07410** [comment]
In reverse order:
- Yes, the copyright is assigned to the project.
- Attribution - sure, I agree but don't know how to do it. :-(    Could you...?
- What's "FileList.txt"???

And please do help with `cryptest.exe tv` stuff (plus, Windows and everything Windows-related is 100% outside my competence).

**Comment by noloader** [comment]
> What's "FileList.txt"???

I use it to build the ZIP. It ensures I don't accidentally include something.

You can `make zip`, then `unzip -aoq cryptopp5.6.3.zip -d /tmp/cryptopp-test`. Then, `cd` into `/tmp/cryptopp-test` and see if things build.

```
$ cat Filelist.txt 
3way.cpp
3way.h
adhoc.cpp.proto
adler32.cpp
adler32.h
aes.h
...
```

**Comment by DevJPM** [comment]
After a quick look through the code, I noticed that you changed some bits with ECIES. (DHAES cofactor option and a change in the comment).

Is this supposed to be in this commit?

**Comment by mouse07410** [comment]
> After a quick look through the code, I noticed that you changed some bits with ECIES. (DHAES cofactor option and a change in the comment). Is this supposed to be in this commit?

Good point. It is one of my personal changes that was necessary to interoperate with BouncyCastle, so strictly speaking it has nothing to do with [F]HMQV.

I will leave it to others to decide whether it's time to make this change official (I'd say the answer is yes). My own fork will keep it regardless, of course (as I need compatibility with BouncyCastle).

**Comment by mouse07410** [comment]
>> What's "FileList.txt"???
>
> I use it to build the ZIP. It ensures I don't accidentally include something.

I see, thanks. Just pushed the addition to `Filelist.txt`.

File: misc.h [hunk 1]
```diff
@@ -66,8 +66,10 @@
 #  define _tzcnt_u32(x) __tzcnt_u32(x)
 #  define _tzcnt_u64(x) __tzcnt_u64(x)
 #  define  _blsr_u32(x)  __blsr_u32(x)
+#ifndef _blsr_u64
 #  define  _blsr_u64(x)  __blsr_u64(x)
 # endif
+# endif
 #endif
```
Comment (noloader) [inline misc.h]: Should this apply to Master now?

Comment (mouse07410) [inline misc.h]: Actually, the same thing should be done for all the 4 macros defined there.

Comment (noloader) [inline misc.h]: OK, can you do that in a different pull request. Guard them all.

**Comment by noloader** [comment]
> was necessary to interoperate with BouncyCastle, 

I think that should probably be separate from this PR.

Can you split the ECIES out into its own PR, and then make the new PR against [Issue 21: ECIES uses byte count, and not bit count, for label size](http://github.com/weidai11/cryptopp/issues/21).

**Comment by mouse07410** [comment]
> I think that should probably be separate PR from this PR.

No argument from me.

> Can you split the ECIES out into its own PR, and then make the new PR against Issue 21: ECIES uses byte count, and not bit count, for label size.

Probably, if somebody could tell me how to remove a commit from one (already existing) PR and place/move it to another (say, new) one.

**Comment by noloader** [comment]
>  if somebody could tell me how to remove a commit from one (already existing) PR and place/move it to another (say, new) one.

Lol... You are asking one of the most git-challenged individuals on the face of the earth :)

I think it will probably be easiest with copy/paste. Perform a fresh checkout, paste the corrected text into this PR, and then push the update to this PR.

**Comment by mouse07410** [comment]
Will try to split this one into three PRs
